### ISSUE TITLE
Weird trigger condition for wrong_self_convention

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust-clippy#9038]
### Description

The following code triggers `wrong_self_convention`
```Rust
mod foo {
    #[derive(Clone, Copy)]
    pub struct Foo<T>(std::marker::PhantomData<T>);

    impl<T> Foo<T> {
        pub fn to_bar(self) -> i32 {
            42
        }
    }
}
```

```
warning: methods with the following characteristics: (`to_*` and `self` type is not `Copy`) usually take `self` by reference
 --> src/lib.rs:6:23
  |
6 |         pub fn to_bar(self) -> i32 {
  |                       ^^^^
```

However, if it's not in a module, or the module is public, the lint doesn't trigger.
   

### Version

```text
rustc 1.63.0-nightly (10f4ce324 2022-06-22)
binary: rustc
commit-hash: 10f4ce324baf7cfb7ce2b2096662b82b79204944
commit-date: 2022-06-22
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.5
```

**Comment by tamaroning** [comment]
This is not a bug. If `fn to_bar` in first comment is exported to other crates, changing its argument causes breakages for other crates. If you want to lint it despite that, please try add `avoid_breaking_exported_api = true` to your clippy.toml.

Ref: https://github.com/rust-lang/rust-clippy/pull/7187

**Comment by crumblingstatue** [comment]
Well, it seems intentional behavior per the above comment, so I'm closing this one.
